FSHSchool/FSHOnline
# Migrate from Create React App to Vite

This PR removes the Create React App framework and migrates to the [Vite](https://vitejs.dev/) framework. This will (hopefully) make maintenance going forward easier to manage with a lighter weight, currently maintained framework. There were quite a few things that changed and I tried to keep them in relatively separate commits, but I've included a brief summary of the changes here:

- Moved `AppRouter` out of `utils` since it didn't really make sense as a util
- Switched all source and test files with React component to be `.jsx` files
- Removed `react-scripts`, `react-app-rewired`, and the serviceWorker code since they are CRA specific
- Added and configured a `vite.config.mjs` file
- Moved `index.html` and updated it for Vite
- Renamed env vars to have VITE_ prefix and used `import.meta` in place of `process.env`
- Stubbed out packages that are file system specific that FSH Online does not need and fhir/fhir because it is used by SUSHI but will never be needed in FSH Online
- Updated ESLint configuration to no longer rely on CRA
- Switched to React 17
- Switched to Node 20
- Switched from Jest (included with react-scripts) to Vitest
- Updated the README to remove references to CRA
- Removed the `bitly` dependency and just make the two calls we need with `fetch` directly

To test this, you should
- `npm install` (it might be good to first delete `node_modules` if you have it to start fresh)
  - I regenerated the `package-lock.json` a few times to make sure it was up to date, so make sure there are no changes after you do an install
- use `npm start` to run the development server
- test the app to make sure it works as expected
  - Try using SUSHI, using GoFSH, loading a Share link, saving FSH/JSON/all files, and using the Examples.
  - Also, remove any saved definitions from your browser's storage in IndexedDB and then try to run SUSHI/GoFSH to ensure packages can still be loaded properly
- use `npm run build` to build a static, production build (this will be in `dist` now, not `build`, since that's the Vite default)
- use `npm run preview` to serve the static build and test this out in the same way as above
- use `npm test` to make sure the tests run and pass
- use `npm run lint` and `npm run prettier` to ensure the updated configuration works as expected

Two other things to note:
- To get the "format on save" to work in VS Code, I did have to add a File Association in my editor settings. Let me know if you run into the same thing and I can help.
- I think I used the `base` property in the `vite.config.mjs` file properly so that the deployment will work, but if anyone notices anything that might specifically be off there, please let me know.
